File: _Demo_and_Tests/Objective-C/AlgorithmsTests/MergeSortTests.m
//
//  MergeSortTests.m
//  Algorithms
//

#import <XCTest/XCTest.h>
#import "MergeSort.h"
#import "TestsHelper.h"

@interface MergeSortTests : XCTestCase

@end

@implementation MergeSortTests

- (void)setUp {
    [super setUp];

    [TestsHelper reset];
}

- (void)tearDown {
    [super tearDown];
}

#pragma mark -

- (void)testAscOdd {
    NSArray *resultArray = [MergeSort sortedArray:TestsHelper.array order:MergeSortOrderAsc];

    XCTAssertEqualObjects(resultArray, TestsHelper.sortedArrayAsc);
}

- (void)testAscEven {
    [TestsHelper.array addObject:@100];
    [TestsHelper.sortedArrayAsc addObject:@100];

    NSArray *resultArray = [MergeSort sortedArray:TestsHelper.array order:MergeSortOrderAsc];

    XCTAssertEqualObjects(resultArray, TestsHelper.sortedArrayAsc);
}

- (void)testDescOdd {
    NSArray *resultArray = [MergeSort sortedArray:TestsHelper.array order:MergeSortOrderDesc];

    XCTAssertEqualObjects(resultArray, TestsHelper.sortedArrayDesc);
}

- (void)testDescEven {
    [TestsHelper.array addObject:@100];
    [TestsHelper.sortedArrayDesc insertObject:@100 atIndex:0];

    NSArray *resultArray = [MergeSort sortedArray:TestsHelper.array order:MergeSortOrderDesc];

    XCTAssertEqualObjects(resultArray, TestsHelper.sortedArrayDesc);
}

- (void)testNilArray {
    NSArray *resultArray = [MergeSort sortedArray:nil order:MergeSortOrderAsc];

    XCTAssert(resultArray == nil);
}

- (void)testEmptyArray {
    NSArray *resultArray = [MergeSort sortedArray:@[] order:MergeSortOrderAsc];

    XCTAssertEqualObjects(resultArray, @[]);
}

- (void)testArrayWithOneNumber {
    NSArray *resultArray = [MergeSort sortedArray:@[@0] order:MergeSortOrderAsc];

    XCTAssertEqualObjects(resultArray, @[@0]);
}

@end

File: _Demo_and_Tests/Swift/AlgorithmsTests/MergeSortTests.swift
//
//  MergeSortTests.swift
//  Algorithms
//

import XCTest

class MergeSortTests: XCTestCase {

    override func setUp() {
        super.setUp()

        TestsHelper.reset()
    }

    override func tearDown() {
        super.tearDown()
    }

    // MARK: -

    func testAscOdd() {
        let resultArray = MergeSort.sort(TestsHelper.array, order: .ASC)

        XCTAssert(resultArray ?= TestsHelper.sortedArrayAsc)
    }

    func testAscEven() {
        TestsHelper.array.append(100)
        TestsHelper.sortedArrayAsc.append(100)

        let resultArray = MergeSort.sort(TestsHelper.array, order: .ASC)

        XCTAssert(resultArray ?= TestsHelper.sortedArrayAsc)
    }

    func testDescOdd() {
        let resultArray = MergeSort.sort(TestsHelper.array, order: .DESC)

        XCTAssert(resultArray ?= TestsHelper.sortedArrayDesc)
    }

    func testDescEven() {
        TestsHelper.array.append(100)
        TestsHelper.sortedArrayDesc.insert(100, at: 0)

        let resultArray = MergeSort.sort(TestsHelper.array, order: .DESC)

        XCTAssert(resultArray ?= TestsHelper.sortedArrayDesc)
    }

    func testNilArray() {
        let nilArray: [Int]? = nil
        let resultArray = MergeSort.sort(nilArray, order: .ASC)

        XCTAssert(resultArray == nil);
    }

    func testEmptyArray() {
        let resultArray = MergeSort.sort([] as [Int], order: .ASC)

        XCTAssert(resultArray ?= []);
    }

    func testArrayWithOneNumber() {
        let resultArray = MergeSort.sort([0], order: .ASC)
        
        XCTAssert(resultArray ?= [0]);
    }

}
